File: js/binaryTree.js
const createNode = (value) => {
  return {value, left: null, right: null};
};

const insert = (tree, value) => {
  if (tree == null) {
    return createNode(value);
  }
  if (value < tree.value) {
    tree.left = insert(tree.left, value);
  } else {
    tree.right = insert(tree.right, value);
  }
  return tree;
};

const printInOrder = (tree) => {
  if (tree == null) {
    return;
  }
  printInOrder(tree.left);
  console.log(tree.value);
  printInOrder(tree.right);
};

const printPreOrder = (tree) => {
  if (tree == null) {
    return;
  }
  console.log(tree.value);
  printPreOrder(tree.left);
  printPreOrder(tree.right);
};

const printPostOrder = (tree) => {
  if (tree == null) {
    return;
  }
  printPostOrder(tree.left);
  printPostOrder(tree.right);
  console.log(tree.value);
};

const search = function (tree, value) {
  if (tree == null) {
    return false;
  }
  if (tree.value == value) {
    return true;
  }
  const remainingTree = tree.value > value ? tree.left : tree.right;
  return search(remainingTree, value);
};

const visitInOrder = function (tree, visiter) {
  let curr = tree;
  const stack = [];
  while (true) {
    while (curr != null) {
      stack.push(curr);
      curr = curr.left;
    }

    if (!stack.length) break;
    curr = stack.pop();
    visiter(curr.value);
    curr = curr.right;
  }
};

const visitPreOrder = function (tree, visiter) {
  let curr = tree;
  const stack = [];
  stack.push(curr);
  while (stack.length) {
    curr = stack.pop();
    visiter(curr.value);
    if (curr.right) {
      stack.push(curr.right);
    }
    if (curr.left) {
      stack.push(curr.left);
    }
  }
};

const visitPostOrder = function (tree, visiter) {
  if (tree == null) return;
  const stack = [];
  do {
    while (tree) {
      if (tree.right) {
        stack.push(tree.right);
      }
      stack.push(tree);
      tree = tree.left;
    }
    tree = stack.pop();
    if (tree.right && stack[stack.length - 1] == tree.right) {
      stack.pop();
      stack.push(tree);
      tree = tree.right;
    } else {
      visiter(tree.value);
      tree = null;
    }
  } while (stack.length);
};

const find_min = function (tree) {
  if (tree.left === null) {
    return tree;
  }
  return find_min(tree.left);
};

const deleteNode = function (tree, value) {
  if (tree == null) {
    return tree;
  }
  if (value < tree.value) {
    tree.left = deleteNode(tree.left, value);
    return tree;
  }
  if (value > tree.value) {
    tree.right = deleteNode(tree.right, value);
    return tree;
  }
  if (tree.left == null) {
    return tree.right;
  }
  if (tree.right == null) {
    return tree.left;
  }
  const minNode = find_min(tree.right);
  tree.value = minNode.value;
  tree.right = deleteNode(tree.right, minNode.value);
  return tree;
};

const calcHeight = function (tree) {
  if (!tree) return 0;
  return 1 + Math.max(calcHeight(tree.left), calcHeight(tree.right));
};

const isBalanced = function (tree) {
  if (tree == null) return true;
  const leftHeight = calcHeight(tree.left);
  const rightHeight = calcHeight(tree.right);
  return Math.abs(leftHeight - rightHeight) <= 1 && isBalanced(tree.left) && isBalanced(tree.right);
};

const leftRotate = function (tree) {
  if (!tree || !tree.right) {
    return tree;
  }
  const pivot = tree.right;
  tree.right = pivot.left;
  pivot.left = tree;
  return pivot;
};

const rightRotate = function (tree) {
  if (!tree || !tree.left) {
    return tree;
  }
  const pivot = tree.left;
  tree.left = pivot.right;
  pivot.right = tree;
  return pivot;
};

const rotate = function (tree, pivot) {
  if (tree.right == pivot) {
    tree = leftRotate(tree);
  }
  if (tree.left == pivot) {
    tree = rightRotate(tree);
  }
  return tree;
};

const balanceTree = function (nodeToBeRoot, tree) {
  if (tree.value === nodeToBeRoot) {
    return tree;
  }
  if (tree.left && tree.left.value === nodeToBeRoot) {
    return rotate(tree, tree.left);
  }
  if (tree.right && tree.right.value === nodeToBeRoot) {
    return rotate(tree, tree.right);
  }
  if (nodeToBeRoot < tree.value) {
    tree.left = balanceTree(nodeToBeRoot, tree.left);
  } else {
    tree.right = balanceTree(nodeToBeRoot, tree.right);
  }
  return balanceTree(nodeToBeRoot, tree);
};

const balance = function (tree) {
  if (isBalanced(tree)) {
    return tree;
  }
  const nodes = [];
  visitInOrder(tree, nodes.push.bind(nodes));
  const nodeToBeRoot = nodes[Math.ceil(nodes.length / 2) - 1];
  tree = balanceTree(nodeToBeRoot, tree);
  tree.left = balance(tree.left);
  tree.right = balance(tree.right);
  return tree;
};

module.exports = {
  balance,
  deleteNode,
  insert,
  visitInOrder,
  visitPreOrder,
  visitPostOrder,
  rotate,
};

File: C/binary_tree.c
#include "binary_tree.h"

Tree_ptr create_tree(int value){
  Tree_ptr tree = (Tree_ptr) malloc(sizeof(Tree));
  tree->value = value;
  tree->left = NULL;
  tree->right = NULL;
  return tree;
}

void print_in_order(Tree_ptr tree) {
  if (tree == NULL) {
    return;
  }
  print_in_order(tree->left);
  printf("%d  ",tree->value);
  print_in_order(tree->right);
};

void print_pre_order(Tree_ptr tree){
  if (tree == NULL) {
    return;
  }
  printf("%d  ",tree->value);
  print_pre_order(tree->left);
  print_pre_order(tree->right);
};

void print_post_order(Tree_ptr tree){
  if (tree == NULL) {
    return;
  }
  print_post_order(tree->left);
  print_post_order(tree->right);
  printf("%d  ",tree->value);
};

Tree_ptr insert(Tree_ptr tree, int value) {
  Tree_ptr_to_ptr current = &tree;
  Tree_ptr temp_node = create_tree(value);

  while(*current != NULL){
    if (value < (*current)->value)
    {
      current = &((*current)->left);
    }else{
      current = &((*current)->right);
    }
  } 
  (*current) = temp_node;
  return tree;
}

Boolean search(Tree_ptr tree, int value){
  Tree_ptr current = tree;
  while(current != NULL){
    if(value == current->value){
      return True;
    }
    current = value < current -> value ? current->left : current->right;
  }
  return False;
}

Tree_ptr find_min(Tree_ptr node)
{ 
  Tree_ptr current = node; 
  while (current->left != NULL) 
    current = current->left; 
  return current; 
} 

Tree_ptr delete_node(Tree_ptr tree, int value)
{
  Tree_ptr_to_ptr current = &tree;
  while (*current && (*current)->value != value)
  {
    current = value < (*current)->value ? &((*current)->left): &((*current)->right);
  } 
  if(*current == NULL)
  { 
    return tree;
  }

  if((*current)->left == NULL)
  {
    (*current) = (*current)->right;
  } else if ((*current)->right == NULL)
  {
    (*current) = (*current)->left;
  } else
  {
    Tree_ptr_to_ptr child = &((*current)->right);
    while ((*child)->left != NULL) { 
      child = &((*child)->left); 
    } 
    (*current)->value = (*child)->value;
    Tree_ptr node_to_delete = *child;
    *child = (*child)->right;
    free(node_to_delete);
  }
  return tree;
}
Tree_ptr left_rotate(Tree_ptr tree) {
  if (!tree || !tree->right) {
    return tree;
  }
  Tree_ptr pivot = tree->right;
  tree->right = pivot->left;
  pivot->left = tree;
  return pivot;
};

Tree_ptr right_rotate(Tree_ptr tree) {
  if (!tree || !tree->left) {
    return tree;
  }
  Tree_ptr pivot = tree->left;
  tree->left = pivot->right;
  pivot->right = tree;
  return pivot;
};

Tree_ptr rotate(Tree_ptr tree, Tree_ptr pivot){
  if(tree == NULL) return tree;
  if(tree->right == pivot){
    tree = left_rotate(tree);
  }
  if(tree->left == pivot){
    tree = right_rotate(tree);
  }
  return tree;
}

int max(int num1, int num2){
  return num1 < num2 ? num2: num1;
}

int calc_height(Tree_ptr tree){
  if(tree == NULL) return 0;
  return 1 + max(calc_height(tree->left), calc_height(tree->right));
}

Boolean is_balanced(Tree_ptr tree){
  if(tree==NULL) return True;
  int left_height = calc_height(tree->left);
  int right_height = calc_height(tree->right);
  return abs(left_height - right_height) <= 1 && is_balanced(tree->left) && is_balanced(tree->right);
}

int store_nodes(Tree_ptr tree, Int_ptr nodes, int no_of_nodes){
  if (tree == NULL)
  {
    return no_of_nodes;
  }
  no_of_nodes = store_nodes(tree->left, nodes, no_of_nodes);
  nodes[no_of_nodes++] = tree->value;
  no_of_nodes = store_nodes(tree->right, nodes, no_of_nodes);
  return no_of_nodes;
}



Tree_ptr balance(Tree_ptr tree) {
  if (is_balanced(tree)) {
    return tree;
  }
  int nodes[100];
  int no_of_nodes =  store_nodes(tree, nodes, 0);
  int node_to_be_root = nodes[(no_of_nodes / 2) - 1];
  while (tree->value != node_to_be_root) {
    tree = rotate_by_value(tree, node_to_be_root);
  }
  tree->left = balance(tree->left);
  tree->right = balance(tree->right);
  return tree;
};

Tree_ptr build_balanced_tree(Int_ptr nodes,int start,int end) {
  if (start > end) {
    return NULL;
  }
  int mid = floor((start + end) / 2);
  Tree_ptr tree = (Tree_ptr)malloc(sizeof(Tree));
  tree->value = nodes[mid];
  tree->left = build_balanced_tree(nodes, start, mid - 1);
  tree->right = build_balanced_tree(nodes, mid + 1, end);
  return tree;
};


Tree_ptr balance_by_creating_new_tree(Tree_ptr tree) {
  int nodes[100];
  int nodes_count = store_nodes(tree,nodes,0);
  return build_balanced_tree(nodes, 0, nodes_count - 1);
};

Tree_ptr rotate_by_value(Tree_ptr tree, int value)
{
 
  Tree_ptr_to_ptr parent = NULL;
  Tree_ptr_to_ptr node = &tree;
  while (*node && value != (*node)->value)
  {
    parent = &(*node);
    node = value < (*node)->value ? &(*node)->left : &(*node)->right;
  }
  if (*node == NULL)
  {
    return tree;
  }
  if (*parent)
  {
    *parent = rotate(*parent, (*node));
  }
  return tree;
};


